composer/utils/checkpoint.py

```diff
@@ -949,16 +967,40 @@ def save_checkpoint(
         _validate_save_planner(save_planner)
 
         import torch.distributed.checkpoint as dist_cp
-
-        log.debug('Saving sharded checkpoints to %s...', save_filename)
-        dist_cp.save_state_dict(
-            state_dict=state_dict,
-            storage_writer=dist_cp.FileSystemWriter(dirname),
-            planner=save_planner,
-        )
+        from torch.distributed import get_process_group_ranks
+
+        log.debug(f'Saving sharded checkpoints to {save_filename}...')
+        process_group = None
+        device_mesh = state.fsdp_device_mesh
+        if device_mesh is not None and device_mesh.ndim == 2:
+            expect_file = (device_mesh.get_local_rank(mesh_dim=0) == 0)
+            process_group = device_mesh.get_group(1)  # Only save on first replica
+            log.debug(
+                f'global_rank={dist.get_global_rank()}, {expect_file=}, process_group={get_process_group_ranks(process_group)}'
+            )
+        else:
+            expect_file = True
```

Comment:
Is 3d device mesh allowed? I'm assuming this else is for ndim=1. If that's correct, let's error for ndim=3

Reply:
It should still work. If we recognize the mesh structure and can save only replica 1, we do so (optimization), but if we don't, we save all replicas (which still works)

Reply:
I don't think FSDP ever would see a 3D device mesh it would only see a 2D one or a 2D submesh with a 3D parent

Reply:
it only sees 1D (fsdp) or 2D (hsdp)